dm/syncer/binlogstream/stream_modifier.go

```diff
@@ -0,0 +1,273 @@
+		if len(events) == 0 {
+			return terror.ErrSyncerEvent.Generatef("%s op should have non-empty events", req.Op.String())
+		}
+	default:
+		m.logger.DPanic("invalid error op", zap.String("op", req.Op.String()))
+		return terror.ErrSyncerEvent.Generatef("invalid error op: %s", req.Op.String())
+	}
+
+	pos, err := binlog.PositionFromStr(req.BinlogPos)
+	if err != nil {
+		return err
+	}
+
+	toInject := newOperator(req.Op, pos, events, req)
+	toInsertIndex := 0
+
+	// not too much op, don't need binary search
+	for i, op := range m.ops {
+		if op.pos.Compare(pos) < 0 {
+			toInsertIndex = i + 1
+		} else {
+			break
+		}
+	}
+
+	if toInsertIndex == len(m.ops) {
+		m.ops = append(m.ops, toInject)
+		return nil
+	}
+
+	pre := m.ops[toInsertIndex]
+	if pre.pos.Compare(pos) == 0 {
+		m.ops[toInsertIndex] = toInject
+		m.logger.Warn("overwrite operator",
+			zap.Stringer("position", pos),
+			zap.Stringer("old operator", pre))
+		return nil
+	}
+
+	m.ops = append(m.ops, nil)
+	copy(m.ops[toInsertIndex+1:], m.ops[toInsertIndex:])
+	m.ops[toInsertIndex] = toInject
+	if toInsertIndex < m.nextOp {
+		m.nextOp++
+	}
+	m.logger.Info("set a new operator",
+		zap.Stringer("position", pos),
+		zap.Stringer("new operator", toInject))
+	return nil
+}
+
+// Delete will delete an operator. `posStr` should be in the format of "binlog-file:pos".
+func (m *streamModifier) Delete(posStr string) error {
+	pos, err := binlog.PositionFromStr(posStr)
+	if err != nil {
+		return err
+	}
+
+	toDeleteIndex := 0
+	// not too much op, don't need binary search
+	for i, op := range m.ops {
+		if op.pos.Compare(pos) < 0 {
+			toDeleteIndex = i + 1
+		} else {
+			break
+		}
+	}
+
+	if toDeleteIndex < m.nextOp || toDeleteIndex == len(m.ops) {
+		return terror.ErrSyncerOperatorNotExist.Generate(posStr)
+	}
+	pre := m.ops[toDeleteIndex]
+	if pre.pos.Compare(pos) != 0 {
+		return terror.ErrSyncerOperatorNotExist.Generate(posStr)
+	}
+	copy(m.ops[toDeleteIndex:], m.ops[toDeleteIndex+1:])
+	m.ops = m.ops[:len(m.ops)-1]
+	return nil
+}
+
+// ListEqualAndAfter returns a JSON string of operators equals and after the given
+// position.
+// - if argument is "", it returns all operators.
+// - Otherwise caller should make sure the argument in format of "binlog-file:pos"
+//   and it returns all operators >= this position.
+func (m *streamModifier) ListEqualAndAfter(posStr string) string {
+	toPrint := []*operator{}
+	if posStr == "" {
+		toPrint = m.ops
+	} else {
+		pos, err := binlog.PositionFromStr(posStr)
+		if err != nil {
+			m.logger.DPanic("invalid position, should be verified in caller",
+				zap.String("position", posStr))
+			return ""
+		}
+
+		i := 0
+		for i < len(m.ops) && m.ops[i].pos.Compare(pos) < 0 {
+			i++
+		}
+
+		if i < len(m.ops) {
+			toPrint = m.ops[i:]
+		}
+	}
+
+	reqs := make([]*pb.HandleWorkerErrorRequest, 0, len(toPrint))
+	for _, op := range toPrint {
+		reqs = append(reqs, op.originReq)
+	}
+
+	b, err2 := json.Marshal(reqs)
```

Comment:
Why is the `reqs` serialized into a JSON string rather than returned directly?

Reply:
We can return reqs and let caller do it. Will change it in this PR later.

Reply:
https://github.com/pingcap/tiflow/pull/5890/commits/69ac38fcf5c4596a458f7bc3f43389cba57d7ba0